### PR TITLE
Java: Add MyBatis Sql Injection no @Param case

### PR DESCRIPTION
**Pull request by retanoj**
If the project is compiled with '-parameters' compiler option, Mybatis mapper function supports no `@Param` annotation parameter which name is the same as it in SQL statement.

In below case,  MyBatisAnnotationSqlInjection should find a vulnerability at name parameter in `bad10` function. That is a FN.

```java
@Select({"select * from test where id = #{id} and name = '${name}'"})
String bad10(Integer id, String name);
```

The same scenario also applies to MyBatisMapperXmlSqlInjection, So I add a snippet check code in `MyBatisCommonLib` for supporting case like this.

### CODE REVIEW

File: java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll [hunk 1]
```diff
@@ -152,6 +153,8 @@ predicate isMybatisXmlOrAnnotationSqlInjection(
         unsafeExpression.matches("${param" + (i + 1) + "%}")
         or
         unsafeExpression.matches("${arg" + i + "%}")
+        or
+        unsafeExpression.matches("${" + ma.getMethod().getParameter(i).getName() + "}")
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll:157]:
```suggestion
        unsafeExpression.matches("${" + ma.getMethod().getParameter(i).getName() + "%}")
```
The `%` operator matches any string, and in this case could consume a modifier like `,jdbcType=VARCHAR` as seen in the example.

I note a small bug in this and the existing cases btw, when a method parameter's name is a prefix of another one, or the same situation for fields. If you had the appetite for it, a good future PR would be to switch to regex matching, and check that the suffix begins with optional whitespace followed by a comma.

Comment (retanoj) [reply java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll:157]: I notice the lack of '%'. My first feeling is "Yes, add it. All of others added it.".
But after test, I noticed Mybatis not support usage like "${var,jdbcType=VARCHAR}". This makes an error occur.
`Parameter 'VARCHAR' not found. Available parameters are [var, param1]`

Comment (smowton) [reply java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll:157]: But you can use `${param1,jdbcType=VARCHAR}`?

Comment (smowton) [reply java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll:157]: Interesting. It's a little concerning considering the `%` on the param case clearly expects that you can... but at the least this query is better than it was, so I'm happy to merge it.

Comment (retanoj) [reply java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll:157]: > But you can use `${param1,jdbcType=VARCHAR}`?

I can use `#{  param1  ,  jdbcType=VARCHAR  }` but not `${param1,jdbcType=VARCHAR}`.
Another small bug , see the space :(

Comment (retanoj) [reply java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll:157]: > Interesting. It's a little concerning considering the `%` on the param case clearly expects that you can... but at the least this query is better than it was, so I'm happy to merge it.

I'm wrong, dude.
On annotation SQL statement, `'${param1,jdbcType=VARCHAR}'` is failed. I do not know the reason.
On XML SQL statement, `'${param1,jdbcType=VARCHAR}'` is working correctly.

File: java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll [hunk 2]
```diff
@@ -152,6 +153,8 @@ predicate isMybatisXmlOrAnnotationSqlInjection(
         unsafeExpression.matches("${param" + (i + 1) + "%}")
         or
         unsafeExpression.matches("${arg" + i + "%}")
+        or
+        unsafeExpression.regexpMatch("\\$\\{\\s*" + ma.getMethod().getParameter(i).getName() + "\\s*(,.*?)?\\s*\\}")
```
Comment (smowton) [inline java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll:157]:
```suggestion
        unsafeExpression.regexpMatch("\\$\\{\\s*" + ma.getMethod().getParameter(i).getName() + "\\s*(,[^}]*)?\\}")
```
Might as well be explicit (untested; I don't think the `}` inside the character class needs escaping though)

Comment (smowton) [reply java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll:157]: If we're doing this here we might as well do it in the other cases too

Comment (retanoj) [reply java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll]: consider blankspace, comma and dot field, I change it to `"\\$\\{\\s*" + var + "\\b[^}]*?\\}"`
case like `${ user . name , jdbcType = VARCHAR , javaType = java.lang.String }`

**Review by smowton — COMMENTED** [review]
All the `*?`s might as well be `*`s considering the follow character is always disjoint with `[^}]` -- otherwise looks good to me!

**Review by smowton — APPROVED** [review]

**Review by smowton — APPROVED** [review]

**Pull request merged**